File: 2360-longest-cycle-in-a-graph/2360-longest-cycle-in-a-graph.py
class Solution:
    def longestCycle(self, edges: List[int]) -> int:
        t = [-1]*len(edges)
        res = [-1]
        vis = set()
        def dfs(cur, i):
            if cur == -1:
                return
            if t[cur] > -1:
                res[0] = max(res[0], i-t[cur])
            else:
                if cur not in vis:
                    vis.add(cur)
                    t[cur] = i
                    dfs(edges[cur], i+1)
                    t[cur] = -1
        for c in range(len(edges)):
            if c not in vis:
                dfs(c, 0)
        return res[0]

File: 2360-longest-cycle-in-a-graph/2360-longest-cycle-in-a-graph.cpp
class Solution {
public:
    set<int> vis;
    unordered_map<int,int> M;
    int dfs(int i, int d, vector<int>& edges) {
        vis.insert(i);
        M[i] = d;
        int res = -1;
        if (edges[i] != -1 && vis.find(edges[i]) == vis.end()) {
            res = dfs(edges[i], d+1, edges);
        }
        else {
            // cout<<i<<" "<<edges[i]<<" "<<M[i]<<" "<<M[edges[i]]<<endl;
            res = M[i] - M[edges[i]] + 1;
        }
        M[i] = 1000000;
        return res;
    }
    
    int longestCycle(vector<int>& edges) {
        int res = 0;
        M[-1] = 1000000;
        for (int i = 0; i < edges.size(); i++) {
            if (vis.find(i) == vis.end())
                res = max(res, dfs(i, 0, edges));
            M[i] = 1000000;
        }
        return res > 0? res:-1;
    }
};